full-qr-powered-lanes-demo/query-rules-lanes/query-rules-lanes.js

```diff
@@ -0,0 +1,68 @@
+const state = {
+  lastLanesReceived: null,
+  lanesIndices: [],
+};
+
+const queryRulesLanes = instantsearch.connectors.connectQueryRules(
+  ({ items, widgetParams, instantSearchInstance }) => {
+    // We don't display anything if we don't receive any lanes from the
+    // Query Rules.
+    if (items.length === 0 || !items[0].lanes) {
+      return;
+    }
+
+    const lanes = items[0].lanes;
+    lanes.sort(function(a, b) {
+      return a.position - b.position;
+    });
+
+    // If the lanes haven't changed after a refinement, we don't need to update
+    // the DOM.
+    if (state.lastLanesReceived === JSON.stringify(lanes)) {
+      return;
+    }
+
+    state.lastLanesReceived = JSON.stringify(lanes);
+    const container = document.querySelector(widgetParams.container);
+
+    // We unmount all previous lanes indices to have an updated InstantSearch
+    // tree.
+    state.lanesIndices.forEach(contentShelvesIndex => {
+      contentShelvesIndex.dispose();
+    });
```

Comment:
@francoischalifour Could you explain to me why here you're doing that instead of

```js
instantSearchInstance.mainIndex.removeWidgets(state.lanesIndices);
```

Reply:
You're right we should rather remove the widgets. Feel free to PR!